src/main/kotlin/com/katanox/tabour/integration/sqs/core/publisher/SqsEventPublisher.kt

```diff
@@ -1,69 +1,124 @@
+import software.amazon.awssdk.services.sqs.model.SendMessageBatchRequestEntry
 
 private val logger = KotlinLogging.logger {}
 
 @Component
 class SqsEventPublisher : IEventPublisherBase {
 
-    @Autowired private lateinit var sqsConfiguration: SqsConfiguration
+    @Autowired
+    private lateinit var sqsConfiguration: SqsConfiguration
 
-    @Autowired private lateinit var tabourConfigs: TabourAutoConfigs
+    @Autowired
+    private lateinit var tabourConfigs: TabourAutoConfigs
 
     override fun getType(): BusType {
         return BusType.SQS
     }
 
     override fun publish(message: String, busUrl: String, messageGroupId: String?) {
-        publish(message, busUrl, SendMessageRequest(), messageGroupId)
+        publishBatch(listOf(message), busUrl, messageGroupId)
+    }
+
+    override fun publishBatch(messages: List<String>, busUrl: String, messagesGroupId: String?) {
+        publish(messages, busUrl, messagesGroupId)
     }
 
     /**
-     * Publishes a message with a pre-configured [SendMessageRequest] which gives you all the options
+     * Publishes a message with [SendMessageBatchRequest] which gives you all the options
      * you may need from the underlying SQS client. Note that the `queueUrl` and `messageBody` must
-     * not be set because they will be set by the this publisher.
+     * not be set because they will be set by the publisher.
      */
     private fun publish(
-        message: String,
+        messages: List<String>,
         busUrl: String,
-        preConfiguredRequest: SendMessageRequest,
-        messageGroupId: String?
+        messageGroupId: String?,
     ) {
-        require(preConfiguredRequest.queueUrl == null) {
-            "attribute queueUrl of pre-configured request must not be set!"
-        }
-        require(preConfiguredRequest.messageBody == null) {
-            "message body of pre-configured request must not be set!"
+        runBlocking {
+            retry(times = tabourConfigs.tabourProperties.maxRetryCount) {
+                doPublish(
+                    messages,
+                    busUrl,
+                    messageGroupId
+                )
+            }
         }
-        val retry = tabourConfigs.retryRegistry().retry("publish")
-        retry.eventPublisher.onError { logger.warn("error publishing message to queue {}", busUrl) }
-        retry.executeRunnable { doPublish(message, busUrl, preConfiguredRequest, messageGroupId) }
     }
 
+    @OptIn(DelicateCoroutinesApi::class)
     private fun doPublish(
-        message: String,
+        messages: List<String>,
+        busUrl: String,
+        messagesGroupId: String?,
+    ) {
+        GlobalScope.launch {
+            messages.chunked(messages.size / sqsConfiguration.sqsProperties.publishingCoroutines).forEach {
+                async {
+                    publishChunk(it, busUrl, messagesGroupId)
+                }.start()
+            }
+        }
+    }
```

Comment:
This is quite a hard to interpret method, even though I have a feeling it might be inspired by whatever messy code I had in SqsBenchmarking 😅 So lets see if I can understand it correctly without running it.
Let's say I pass in 50 messages. line 71 is going to be 
`messages.chunked(50/50).forEach {...}`
which, after checking kotlin docs, will assume you want 50 chunks of size 1 and launch 50 coroutines that each publish a single message.. Surely it's more desirable to launch 5 coroutines that do 5 calls with 10 messages?
Second - what happens if I try to publish a list of 10 messages? When I put this line into kotlin playground
`val chunks = words.chunked(10/50)` I get a run-time error.
Third - does this preserve event ordering? Or is the intention that the order doesn't matter within the batch? Even if the order in the batch doesn't matter, If i call this twice in a row with a single message (aside the fact I think the code might fail with message.size under 50) what's to say that the two coroutines will run in order?
Fourth - what happens if I call the publisher twice in a row with 50 messages each - won't both of them spin up 50 coroutines and exhaust the sqs connection pool?

Reply:
i will agree with mantas, using directly the `publishChunk` method should be enough, since it already chunks 10 per batch, so if you pass 10 messages, you will get 1 batch instead of 10, same with if you pass 1 message

Reply:
For the first and second what you mentioned was valid a more in-depth calculation should be done to decide the optimal chunks.
The third one, With the current implementation we assume that the order of messages is being given at first and we don't preserve the order so that should be fixed.
The fourth is correct as well.
I will redesign the publisher to make it more convenient probably